ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/SqlLedger.scala

```diff
@@ -314,42 +320,105 @@ private final class SqlLedger(
 
   override def allocateParty(
       party: Party,
-      displayName: Option[String]): Future[PartyAllocationResult] =
-    ledgerDao
+      displayName: Option[String],
+      submissionId: SubmissionId,
+      participantId: ParticipantId): Future[SubmissionResult] = {
+
+    val storePartyAllocationEntry: Future[SubmissionResult] = {
+      var headRef = 0L
+      ledgerDao
+        .storePartyAllocationEntry(
+          headRef,
+          headRef + 1,
+          None,
+          submissionId,
+          participantId,
+          //TODO BH proper participant isLocal check needed
+          PartyAllocationLedgerEntry
+            .Accepted(
+              submissionId,
+              participantId,
+              timeProvider.getCurrentTime,
+              PartyDetails(party, displayName, isLocal = true))
+        )
+        .map {
+          case PersistenceResponse.Ok =>
+            SubmissionResult.Acknowledged
+          case PersistenceResponse.Duplicate =>
+            SubmissionResult.Acknowledged
+        }(DEC)
+    }
+
+    val storeParty = ledgerDao
       .storeParty(party, displayName, None)
       .map {
         case PersistenceResponse.Ok =>
-          PartyAllocationResult.Ok(PartyDetails(party, displayName, true))
+          SubmissionResult.Acknowledged
         case PersistenceResponse.Duplicate =>
-          PartyAllocationResult.AlreadyExists
+          SubmissionResult.Acknowledged
       }(DEC)
 
+    storePartyAllocationEntry.flatMap {
+      case SubmissionResult.Acknowledged =>
+        storeParty
+      case _ =>
+        Future.failed(ErrorFactories.invalidArgument("unable to store party to DB"))
+    }(DEC)
+  }
+
   override def uploadPackages(
       knownSince: Instant,
       sourceDescription: Option[String],
-      payload: List[Archive]): Future[UploadPackagesResult] = {
-    val submissionId = UUID.randomUUID().toString
+      payload: List[Archive],
+      submissionId: SubmissionId,
+      participantId: ParticipantId): Future[SubmissionResult] = {
+
     val packages = payload.map(archive =>
       (archive, PackageDetails(archive.getPayload.size().toLong, knownSince, sourceDescription)))
-    ledgerDao
-      .uploadLfPackages(submissionId, packages, None)
-      .map { result =>
-        result.get(PersistenceResponse.Ok).fold(logger.info(s"No package uploaded")) { uploaded =>
-          logger.info(s"Successfully uploaded $uploaded packages")
-        }
-        for (duplicates <- result.get(PersistenceResponse.Duplicate)) {
-          logger.info(s"$duplicates packages discarded as duplicates")
-        }
-        // Unlike the data access layer, the API has no concept of duplicates, so we
-        // discard the information; package upload is idempotent, apart from the fact
-        // that we only keep the knownSince and sourceDescription of the first upload.
-        UploadPackagesResult.Ok
-      }(DEC)
+
+    def storePackageUploadEntry(offsets: SqlLedger.Offsets) = {
+      ledgerDao
```

Comment:
Now the storing of the package and the storing of the packageUploadEntry happen in a different database transaction, but this should rather happen atomically. See `JdbcLedgerDao#storeConfigurationEntry`.

Reply:
these are separate Update message though -- if there is a DAR with 3 packages there will be 3 PublicPackageUploaded messages (one for each package) + a 4th message for the accept of the lot.  In the case of any of the packages being rejected there will be 0 PublicPackageUploaded and 1 PackageUploadRejected

submissionId is currently the common identifier across these messages but you don't know when you receive the first PublicPackageUploaded how many more are to come and when the Accept/Reject will come.

When each of the entries are processed in kv land they will call logEntryToUpdate in KeyValueConsumption and fire the updates

We could potentially wrap all the PublicPackageUploaded in the PackageUploadAccepted message and process them as a batch to get atomicity

Reply:
@gerolf-da can i request that we track this as a follow-on PR as it will have some significant knock-on effect to bundle these together and guarantee atomicity